packages/eslint-plugin/tests/rules/no-unnecessary-type-parameters.test.ts

```diff
@@ -0,0 +1,485 @@
+        {
+          messageId: 'sole',
+          data: { name: 'T' },
+          line: 6,
+        },
+      ],
+    },
+    {
+      // See https://github.com/cartant/eslint-plugin-etc/issues/15
+      // and https://stackoverflow.com/a/62093430/388951
+      // See above for a fixed version.
+      code: `
+        /**
+         * Call two functions with the same args, e.g.
+         *
+         * onClick={both(action('click'), setState)}
+         */
+        export function both<
+          Args extends unknown[],
+          CB1 extends (...args: Args) => void,
+          CB2 extends (...args: Args) => void,
+        >(fn1: CB1, fn2: CB2): (...args: Args) => void {
+          return function (...args: Args) {
+            fn1(...args);
+            fn2(...args);
+          };
+        }
+      `,
+      errors: [
+        { messageId: 'sole', data: { name: 'CB1' } },
+        { messageId: 'sole', data: { name: 'CB2' } },
+      ],
+    },
+
+    // Tests from DefinitelyTyped-tools / eslint-plugin / no-unnecessary-generics
+    {
+      code: `
+        type Fn = <T>() => T;
+      `,
+      errors: [{ messageId: 'sole' }],
+    },
+    {
+      code: `
+        type Ctr = new <T>() => T;
+      `,
+      errors: [{ messageId: 'sole' }],
+    },
+    {
+      code: `
+        const f1 = <T,>(): T => {};
+      `,
+      errors: [{ messageId: 'sole' }],
+    },
+    // This is an error: Type parameters cannot appear on a constructor declaration.
+    // `
+    //   class C {
+    //     constructor<T>(x: T) {}
+    //   }
+    // `,
+    {
+      code: `
+        function f2<T>(): T {}
+      `,
+      errors: [{ messageId: 'sole' }],
+    },
+    // The second "T" here is a property name, not a type.
+    // So T is completely unused and this should be caught by no-unused-variables.
+    // `
+    //   function f3<T>(x: { T: number }): void;
+    // `,
+    {
+      code: `
+        function f4<T, U extends T>(u: U): U;
+      `,
+      errors: [{ messageId: 'sole' }],
+    },
+    {
+      code: `
+        const f5 = function <T>(): T {};
+      `,
+      errors: [{ messageId: 'sole' }],
+    },
+    {
+      code: `
+        interface I {
+          <T>(value: T): void;
+        }
+      `,
+      errors: [{ messageId: 'sole' }],
+    },
+    {
+      code: `
+        interface I {
+          m<T>(x: T): void;
+        }
+      `,
+      errors: [{ messageId: 'sole' }],
+    },
+
+    // These tests are new for the no-unnecessary-type-parameters rule
+    {
+      code: `
+        // The inferred return type is Map<V, V>, but these are the sole uses
+        // of both type parameters, so this is invalid.
+        function makeMap<K, V>() {
+          return new Map<K, V>();
+        }
+      `,
+      errors: [
+        { messageId: 'sole', data: { name: 'K' } },
+        { messageId: 'sole', data: { name: 'V' } },
+      ],
+    },
+    {
+      code: `
+        // Same as the previous test, but with an explicit return type.
```

Comment:
[Style] Nit: we generally don't put many comments in tests that can be inferred from the tests themselves. They have a bad habit of getting out of date over time.

```suggestion
```

Here and other ones, per your discretion of what's useful. Non-blocking from my end.